### PR TITLE
implement listener tree

### PR DESCRIPTION
**Pull request by malash**
# Features

1. `ListenerTree` fully support `**` listen in `eventemitter2` so is time to remove `eventemitter2`.

2. `ListenerTree` exports `on`, `off`, `emit` methods. The usage is different from normal 
`EventEmitter`.
```js
const event = new ListenerTree();
event.on(['a', 'b'], callback);
event.on(['a', 'c'], callback);
event.emit(['a']); // callback will be called once
event.off(callback); // will remove all listener of callback
```
3. `ListenerTree` has some algorithm improvements.
3.1 `O(n)` one-path traversal when `on` called. `n` is deep of listening path.
3.2 `O(m)` one-path traversal when `emit` called will run faster if `m` (`emit` path deep) is less than `n` (`on` path deep). It will be useful for change of top state.
3.3 `O(l)` traversal when `off` called because it still need to traverse entire tree. `l` is number of listeners.
```js
event.on(['a', 'b', 'c', 'd', 'e', 'f'], callback); // traverse to 'f'
event.on(['a', 'g', 'h', 'i', 'j', 'k', callback); // traverse to 'k'
event.emit(['a']); // will only traverse to 'a'
event.emit([]); // will only traverse entire tree, that means 'a.b...f' and 'a.g...k'
```

~4. Fix normal bug in `EventEmitter`: listener calls even if it was removed during `emit` processing.~ The 'bug' still exists.

# Breaking changes
1. As a result of feature#2, `state.off` and `state.removeEventListener` only accept on argument `callback`.

# Non-breaking changes

1. The `state.on(...)()` usage keeps compatible so **ALL** `@noflux/react` test passed :)

image

# Performance

benchmark result:

```
ListenersTree x 6,817 ops/sec ±1.96% (75 runs sampled)
HashListenerTree x 2,202 ops/sec ±3.44% (72 runs sampled)
Fastest is ListenersTree
```

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -71,3 +73,5 @@ export const setByPath = (obj, path = [], value) => {
 export const arrayFromAllowNullOrUndefined = arrayLike => (
   isNullOrUndefined(arrayLike) ? [] : [...arrayLike]
 );
+
+export const getRandomId = () => Math.random().toString(16).slice(2);
```
Comment (ssnau) [inline src/utils.js]: can we change it to `getNextId`? 

- If id is incremental, it is easier for us to tell which listener bind first and which comes later? 
- no way to conflict.

File: src/listener-tree.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+          delete node.children[child];
+        }
+      }
+      if (!hasSubtreeListener) {
+        delete node.subtreeListeners[listenerId];
+      }
+    }
+  }
+
+  off(listener) {
```
Comment (ssnau) [inline src/listener-tree.js]: on和off看上去不是对称的。Technically, `on(path, fn)` 应该对应上`off(path, fn)`.

Comment (malash) [reply src/listener-tree.js]: I refactor the listener tree by using recursion and now `on(path, cb)`, `off(path, cb)`, `offAll(path, cb)`, `on => handler` and `emit(path)` work as expected.

File: src/listener-tree.js [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  }
+
+  addListener(path, listener) {
+    const listenerId = getListenerId(listener, true);
+    this.__traverse({
+      path,
+      createEmptyPath: true,
+      callbackAfterRecursion: node => node.updateSubtreeListeners(),
+      callbackAtBottom: node => {
+        if (node.ownListenersTime[listenerId] > 0) {
```
Comment (ssnau) [inline src/listener-tree.js]: `ownListenerCount`?

Comment (malash) [reply src/listener-tree.js]: fixed

**Comment by ssnau** [comment]
LGTM

**Pull request merged**